### DISCUSSION TITLE
Alternative IDE-s

### DISCUSSION THREAD

**Discussion by aivarannamaa** [pybricks/support#707] [General]
Are there any known alternative IDE-s (beside code.pybrics.com and beta.pybricks.com) for developing with Pybricks 3.0 (i.e. via Bluetooth, not via SSH)?

I've got the impression, that pybricksdev is something that could be handy in creating such an alternative IDE, either directly or as a substitute for protocol documentation. Is this correct?

I'm just wondering what would it take to add support for Pybricks to [Thonny IDE](https://thonny.org).

**Comment by dlech** [comment]
Pybricks Code is the only officially supported IDE for Powered Up hubs. Everything is open source, so we welcome any community contributions to support other IDEs. And yes pybricksdev is a good place to start to learn how to download and run a program and can be used directly as a library if you are already using Python.

**Reply by aivarannamaa** [reply]
Thank you for the reply!

I noticed that `Pybricks MicroPython v3.2.0b3 on 2022-07-20; SPIKE Prime Hub with STM32F413VG` lacks `os` / `uos` modules. Is this a temporary situation? How can I perform file operations other than upload/download (eg. listing the files and directories)?

**Reply by laurensvalk** [reply]
Thanks for reaching out @aivarannamaa. Nice to hear that you are trying out Pybricks!

With Pybricks, the hubs work a bit differently than the typical MicroPython board. Most of the hubs have no dedicated storage, so there is no file system at all. This is currently also true for the SPIKE hubs, even those could support a file system (eventually).

Instead, there is a download-and-run operation that will send one precompiled `.mpy` data blob to the hub where it is then executed.

Pybricksdev contains one way of doing this, which uses Bleak for the BLE connection.

We are currently adjusting the download process so that the (one) `.mpy` file remains stored on the hub even after shutdown.

**Reply by aivarannamaa** [reply]
I see. But there will be only one "slot" for stored program, right?

And in terms of the source code, the reuse of code needs to be done by copy-and-paste, because there is no way for the users to import their own shared modules, right?

**Reply by laurensvalk** [reply]
Re-usable code with modules is coming in the near future: https://github.com/pybricks/pybricks-micropython/pull/115

We have some ideas about slots: https://github.com/pybricks/support/issues/139, but for the moment there will be only 1. I would be interested to hear your thoughts about more slots. Would this be used a lot in practice?

But with modules supported, it is already possible for users to make a "main menu", and then import a "slot_x" module to mimic slot behavior.

**Reply by aivarannamaa** [reply]
> I would be interested to hear your thoughts about more slots. Would this be used a lot in practice?

Actually, I'm not part of robotics community myself, so I don't know. I think I would prefer the simplest solution i.e. without a compulsory main menu for choosing the active slot. As you said -- the users can implement their own menu if needed, even when only single module is supported.
